File: Maximum_Number_of_Events_That_Can_Be_Attended_II.java
class Solution {
    public int maxValue(int[][] events, int k) {
        int n = events.length;
        Arrays.sort(events, (a, b) -> (a[1] - b[1]));
        TreeMap<Integer, Integer> dp = new TreeMap<>();
        TreeMap<Integer, Integer> dp2 = new TreeMap<>();
        dp.put(0, 0);
        dp2.put(0, 0);
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < n; j++) {
                int cur = dp.lowerEntry(events[j][0]).getValue();
                if (cur + events[j][2] > dp2.lastEntry().getValue()) {
                    dp2.put(events[j][1], cur + events[j][2]);
                }
            }
            dp = dp2;
            dp2 = new TreeMap<>();
            dp2.put(0, 0);
        }
        return dp.lastEntry().getValue(); 
    }
}

File: Maximum_Number_of_Events_That_Can_Be_Attended_II.py
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        events.sort(key=lambda x: x[1])
        cur, temp = [[0, 0]], [[0, 0]]

        for _ in range(k):
            for start, end, value in events:
                i = bisect.bisect(cur, [start]) - 1
                if cur[i][1] + value > temp[-1][1]:
                    temp.append([end, cur[i][1] + value])
            
            cur, temp = temp, [[0, 0]]
        
        return cur[-1][-1]
